batch/batch/cloud/gcp/driver/zones.py

```diff
@@ -104,11 +109,11 @@ def compute_zone_weights(
         local_ssd_data_disk: bool,
         data_disk_size_gb: int,
         preemptible: bool,
-        region: Optional[str],
+        regions: Optional[List[str]],
```

Comment:
One is that we don't have a preference for regions so include all of them (`regions=None`). The latter is that I care about which regions and there are no regions that are okay to include (`regions=[]`). I think we should allow `None`, but assert the list if given can't be empty.

Reply:
Hm, I still feel like this is very error prone. Can I propose a different way to frame the `regions` variable? Instead of saying that `regions` is list of regions the job *can* run in, `regions` is a set of regions where the job is *required* to run in one of them. An empty set means there are no requirements, while a set with elements indicates that there are restrictions on where a job is allowed to run.

Reply:
Something I just noticed: what happens if a user writes `j.regions([])`?

Reply:
I still find your way confusing. Can you explain in words what you think these cases mean? And rewrite the code statements below to be correct with your interpretation?

`regions = []`
`regions = ['us-central1']`
`regions = ['us-central1', 'us-east1']`

Here's the line where they are used:

```python3
        weights = []
        for region_name, r in self._region_info.items():
            if regions is not None and region_name not in regions:
                continue

            quota_remaining = {q['metric']: q['limit'] - q['usage'] for q in r['quotas']}
```

```python3
    def get_instance(self, user: str, cores_mcpu: int, regions: List[str]):
        i = self.healthy_instances_by_free_cores.bisect_key_left(cores_mcpu)
        while i < len(self.healthy_instances_by_free_cores):
            instance = self.healthy_instances_by_free_cores[i]
            assert cores_mcpu <= instance.free_cores_mcpu
            if user == 'ci' and instance.region == self._ci_region:
                return instance
            if user != 'ci' and (regions is None or (regions and instance.region in regions)):
                return instance
            i += 1
        return None
```

Reply:
At the end of the day I think both of our views work out similarly in practice, and we find each other's preference equally confusing so there probably isn't a right answer here. If you think my interpretation is too confusing then let's go with your approach. I just have a strong tendency to avoid requirements on data that are not expressible in the type system. Here's my interpretation:


`regions = []` = there are no regions the job is required to run in
`regions = ['us-central1']` = the job is required to run in `us-central1`
`regions = ['us-central1', 'us-east1']` = the job is required to run in either `us-central1` or `us-east1`

The code blocks are very similar:

```python
        weights = []
        for region_name, r in self._region_info.items():
            if len(regions) > 0 and region_name not in regions:
                continue

            quota_remaining = {q['metric']: q['limit'] - q['usage'] for q in r['quotas']}
```

```python
    def get_instance(self, user: str, cores_mcpu: int, regions: List[str]):  # Yours needs to be Optional[List[str]] to satisfy mypy
        i = self.healthy_instances_by_free_cores.bisect_key_left(cores_mcpu)
        while i < len(self.healthy_instances_by_free_cores):
            instance = self.healthy_instances_by_free_cores[i]
            assert cores_mcpu <= instance.free_cores_mcpu
            if user == 'ci' and instance.region == self._ci_region:
                return instance
            if user != 'ci' and (len(regions) == 0 or instance.region in regions):
                return instance
            i += 1
        return None
```

Reply:
I fixed this by converting the JSON_ARRAYAGG result being None to be filled in with all possible regions. I think this satisfies both of us.